### PR TITLE
feat: added block hashing

### PR DESCRIPTION
**Pull request by Mirko-von-Leipzig**
I'm not totally stoked about the naming of the hash function - `double_hash`.

### CODE REVIEW

File: src/protocol/payload/mod.rs [hunk 1]
```diff
@@ -126,10 +126,14 @@ impl VarStr {
     }
 }
 
-#[derive(Debug, PartialEq)]
-struct Hash([u8; 32]);
+#[derive(Debug, PartialEq, Clone)]
```
Comment (niklaslong) [inline src/protocol/payload/mod.rs:129]: Is `Clone` here necessary?

File: src/protocol/payload/block.rs [hunk 2]
```diff
@@ -176,10 +174,42 @@ impl Header {
...
+        buffer.write_all(&self.nonce)?;
+
+        self.solution_size.encode(buffer)?;
+        buffer.write_all(&self.solution)?;
+
+        Ok(())
+    }
+
+    /// Calculates the double Sha256 hash for [Header]
+    fn double_hash(&self) -> std::io::Result<Hash> {
```
Comment (niklaslong) [inline src/protocol/payload/block.rs:195]: What about `double_sha256`?

File: src/protocol/payload/block.rs [hunk 3]
```diff
@@ -2,6 +2,10 @@ use crate::protocol::payload::{read_n_bytes, Hash, ProtocolVersion, Tx, VarInt};
 
 use std::io::{self, Cursor, Write};
 
+use sha2::Digest;
+
+use std::convert::TryInto;
```
Comment (niklaslong) [inline src/protocol/payload/block.rs:7]: I'd consolidate this with the `use std...` above ☝️

**Review by niklaslong — COMMENTED** [review]
LGTM with a few comments 👍

File: src/protocol/payload/mod.rs (hunk 1 above)
Comment (Mirko-von-Leipzig) [reply src/protocol/payload/mod.rs:129]: Ah its not currently. I had use for it temporarily, will remove.

**Review by niklaslong — APPROVED** [review]

**Pull request merged**